projects/packages/publicize/src/social-image-generator/class-extractor.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Extractor class.
+ *
+ * @package automattic/jetpack-publicize
+ */
+
+namespace Automattic\Jetpack\Publicize\Social_Image_Generator;
+
+/**
+ * Class for extracting SIG-related settings for a post.
+ */
+class Extractor {
+	/**
+	 * Post to get information from.
+	 *
+	 * @var int $post_id
+	 */
+	public $post_id;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param int $post_id Post to get information from.
+	 */
+	public function __construct( $post_id ) {
+		$this->post_id = $post_id;
+	}
+
+	/**
+	 * Get the SIG settings for a post.
+	 *
+	 * @return array
+	 */
+	private function get_post_settings() {
+		$social_options = get_post_meta( $this->post_id, \Automattic\Jetpack\Publicize\Publicize::POST_JETPACK_SOCIAL_OPTIONS, true );
+
+		if ( ! is_array( $social_options ) || empty( $social_options['image_generator_settings'] ) ) {
+			return array();
+		}
+
+		return $social_options['image_generator_settings'];
+	}
+
+	/**
+	 * Check if SIG is enabled for a specific post.
+	 *
+	 * @return bool
+	 */
+	public function is_enabled() {
+		$settings = $this->get_post_settings();
+
+		return ! empty( $settings['is_enabled'] );
+	}
+
+	/**
+	 * Get the text to use for the generated image.
+	 *
+	 * @return string
+	 */
+	public function get_generated_image_text() {
+		$settings = $this->get_post_settings();
+
+		if ( ! empty( $settings['custom_text'] ) ) {
+			return $settings['custom_text'];
+		}
+
+		return get_the_title( $this->post_id );
+	}
+
+	/**
+	 * Get the background image to use for the generated image.
+	 *
+	 * @return string
+	 */
+	public function get_generated_image_background_image_url() {
+		$settings = $this->get_post_settings();
+
+		if ( empty( $settings['image_id'] ) ) {
+			return '';
+		}
+
+		$image = wp_get_attachment_image_url( $settings['image_id'] );
```

Comment:
The default size here is thumbnail, so it makes the image pretty grainy.

![image](https://user-images.githubusercontent.com/96462/221319876-153d75b7-6fc0-467c-9887-f132d86675bf.png)

Reply:
Whoops.. fixed in https://github.com/Automattic/jetpack/pull/29118/commits/0115dc04e0e813bd846910f80fae5627f78f8c54.

Reply:
I think it is. It'll be slightly upscaled, but I don't think it'll be noticeable since these images are viewed at a much smaller size 99% of the time.

The alternatives I see are 1) adding our own custom image size or 2) defaulting to `full`, which has the potential to send over some gigantic images. I don't love either of those options—do you know if there's a third option I'm missing?

Reply:
Yeah I think I'd go with 1 over 2, especially as we load the image into memory to generate the OG image. People are a bit picky about images, but large is probably fine for now.